### DISCUSSION TITLE
Version 0.29 Alpha

### DISCUSSION THREAD

**Discussion by riclolsen** [riclolsen/json-scada#114] [General]
Notes for version 0.29:

* Default login credentials: username=admin password=jsonscada.
* New IEC61850 Client protocol driver.
* C# MongoDB drivers updated.
* Npm packages updated.
* Refactored Nodejs code for updated mongodb driver.
* MongoDB Compass updated to 1.39.0.
* MongoDB Community Server updated to 6.0.8.
* Grafana updated to 9.5.6.
* Chromium updated to 115.0.5790.102.
* Telegraf runtime updated to 1.27.2.
* Dotnet runtime updated to 6.0.19.
* Nodejs runtime updated to 20.5.0.

Requirements and recommendations:

* Windows 10/11 64 bits or Server >=2016, Windows PowerShell. At least 10GB of free space in the "C:" drive.
* Administrative rights. Corporative Windows policies may cause problems with the creation of services and the opening of TCP ports.
* Free TCP ports 27017, 5432, 80, 8080, 3000, 9000. Other ports may be required for optional services and protocols.
* If the server already has MongoDB, PostgreSQL, Grafana, Nginx or another webserver, please uninstall, disable or watch out for possible conflicts.
* Do not update previously installed JSON-SCADA. Please uninstall previous JSON-SCADA versions before installing a new version.

<hr /><em>This discussion was created from the release <a href='https://github.com/riclolsen/json-scada/releases/tag/V0.29-alpha'>Version 0.29 Alpha</a>.</em>

**Comment by arifmjcet** [comment]
Hi Ricardo, I am testing json-scada 0.29 on IEC 61850 ieds, Iam facing following issues.
 
1)I am unable to use excel based configuration tool. Iam getting error( "runtime error:438: Object doesn't support this property or method"). I ensured that "config_server_excel" windows service was running.

 2)How to add parameters related to protocol destination for a tag? In web based configuration of tags, parameters related to protocol source are only available to configure. My objective is to get data from IEC 61850 servers using IEC 61850 MMS client protocol driver and then send the same to a IEC-104 client using IEC-104 server protocol driver.

  By the way Great work.

**Reply by riclolsen** [reply]
Hi,
1) I will check this problem.
2) The web UI is not complete in this regard. You can try using MongoDB Compass and update the tags on _protocolDestinations_ array as described here https://github.com/riclolsen/json-scada/blob/master/src/lib60870.netcore/iec104server/README.md#configure-tags-for-distribution.

**Reply by riclolsen** [reply]
I have tested the Excel config and it works with Excel 2016 and the current 365 version.
The problem can be an older version of Excel or perhaps some locale incompatibility. Please try to use EN-US locale if possible.
Also if you can debug the error, please inform the line of code that is causing the error.

**Reply by arifmjcet** [reply]
Hi, I have tested excel based configuration using Excel 2021, it is working perfectly fine. It is not working with older versions of excel. Thanks for the support.

**Comment by riclolsen** [comment]
Thanks, this is important to know.

**Comment by arifmjcet** [comment]
Hi Ricardo, while trying to use json scada as iec 61850 client, tags are not getting updated by MMS reports from server. Every time json scada client gets a report, i am getting an error ("Value cannot be null. (Parameter 'source')") in the log. Although data is correctly updating using polling. I checked the server side, it is sending reports correctly. I have attached the log file, any help is much appreciated.  
[61850log.txt](https://github.com/riclolsen/json-scada/files/12327294/61850log.txt)

**Reply by riclolsen** [reply]
Hi.
The log does not include the error.

**Reply by arifmjcet** [reply]
I am getting  this line in the log " value cannot be null.( Parameter 'source')" every time server sends a report to json scada client. I have enabled urcbD01 report only, by adding a topic  in web based config. In log, all other rcbs except urcbD01 are de activated. Even on server UrcbD01 is showing as enabled and reserved.

**Reply by riclolsen** [reply]
Can you please post the log with the error?

**Reply by arifmjcet** [reply]
I have uploaded the log. If you open the log with notepad ++,  line No 334 to 356 has the line "Value cannot be null. (Parameter 'source')". This line is getting logged every time server is sending a report( I have configured report integrity to 5 sec by configuring class0ScanInterval to 5, via web based config tool). I was able to monitor the log by double clicking iec61850_client.exe file in bin folder.
[61850lognew.txt](https://github.com/riclolsen/json-scada/files/12327618/61850lognew.txt)

**Reply by riclolsen** [reply]
Please extract to c:\json-scada\bin and retry.
[iec61850_client.zip](https://github.com/riclolsen/json-scada/files/12328080/iec61850_client.zip)

**Reply by arifmjcet** [reply]
Hi, i have extracted the zip file into the c:\json-scada\bin folder and checked if the issue resolves. Now, every time the server sends the report, "Object reference not set to an instance of an object." is logged in the json scada client. I have posted the log for your kind reference( notepad++ line number 274 onwards).
[61850log13082023.txt](https://github.com/riclolsen/json-scada/files/12329831/61850log13082023.txt)

**Reply by riclolsen** [reply]
Ok, please try again with this new binary and post the new log. This just will produce a more detailed error log, helping me to locate the problem.

[iec61850_client.zip](https://github.com/riclolsen/json-scada/files/12330307/iec61850_client.zip)

**Reply by arifmjcet** [reply]
Hi, i have posted the new log for your reference(notepad++ line 284 onwards).
[61850log14082023.txt](https://github.com/riclolsen/json-scada/files/12331612/61850log14082023.txt)

**Reply by riclolsen** [reply]
I think I understand the problem now.
Reports does not contain data references. This is not implemented yet in this driver. 
Processing reports without references will require more time to implement, now I see it is necessary.
Let's see if it works by setting optional bit for requesting data references be included in reports.

Please try this new version, post the logs.

[iec61850_client.zip](https://github.com/riclolsen/json-scada/files/12340190/iec61850_client.zip)

**Reply by arifmjcet** [reply]
I have tried the new version and I have posted the new log( Notepad++ line 279 onwards). Now, the json-scada client is receiving the reports sent by the server. One thing I have observed is, every time a report is received, new tags are getting created and the tags which are linked to the received data are also getting updated with the new data. I think only the tags which are associated with the data(using protocol source object address) should be updated.  Also, "auto create tags" setting is disabled. Thanks for the great work!
[61850log15082023.txt](https://github.com/riclolsen/json-scada/files/12344110/61850log15082023.txt)

**Reply by riclolsen** [reply]
Thanks. You can try this new update to see if it can avoid creating new tags.

[iec61850_client.zip](https://github.com/riclolsen/json-scada/files/12353489/iec61850_client.zip)

**Reply by arifmjcet** [reply]
I have tried the new version, it is working perfectly fine for analog data. Thank you very much. I would like to report one more issue I am facing with the json scada client. For digital tags(double point status) , data is not read properly from the server. Example: for status= 0,1 it is showing v=0, but for status= 2, and 3 the value is read as v=1  (61850log1 notepad++ line 272, 277). Also, json scada client is not adding digital status as a "desired entry" when protocol source object address is set as: R2L6M1Control/XCBR1.Pos.stVal (full path) (61850log1 notepad++ line 102)  and reports are not getting logged for digital status. When I set protocol source object address as: R2L6M1Control/XCBR1.Pos, reports are getting logged but correct value is not updated in digital tag, for all possible states, v=0 is logged. MMS bitstring is reflecting correct value(61850log2 notepad++ line 314). I have posted both the logs  when the actual value of digital tag is =2(ON status).
[61850log1.txt](https://github.com/riclolsen/json-scada/files/12360403/61850log1.txt) 
[61850log2.txt](https://github.com/riclolsen/json-scada/files/12360416/61850log2.txt)

**Reply by arifmjcet** [reply]
One more thing I have observed is,  if full path is configured in the protocol source object address field in the digital status tag configuration  (eg:  R2L6M1Control/XCBR1.Pos.stVal), Although I am getting a value 0( for 00 and 01) or 1 ( for 10 and 11), in HMI graphics, it is correctly showing the digital double point status. On more thing I would like to report is, if I distribute this digital status  tag to an iec- 104 client using the json scada IEC- 104 server driver ,  the value of the digital state is correctly  sent for two states 01 and 10. Only reports are not working.

**Comment by arifmjcet** [comment]
Hi Ricardo, While using excel based configuration tool to configure Iec 61850 tags, the protocol common address is of type "text" and expected entry in the cell is of type int.64. I was able to get around this problem by changing following in realtimeData query in excel file from {"protocolSourceCommonAddress",  Int64.Type} to {"protocolSourceCommonAddress", type text} and also replacing following line,
 elem.protocolSourceCommonAddress = toMongoDouble(elem.protocolSourceCommonAddress ) with elem.protocolSourceCommonAddress = toMongoDoubleOrString(
        elem.protocolSourceCommonAddress
      ),  in index.js file in C:\json-scada\src\config_server_for_excel.

**Reply by riclolsen** [reply]
There is indeed a problem here. I will fix that. Thanks for reporting.

**Reply by riclolsen** [reply]
Fixed in the source code.

**Reply by arifmjcet** [reply]
Thanks!
